source/server/admin/grouped_stats_request.cc

```diff
@@ -0,0 +1,156 @@
+}
+
+template <class SharedStatType>
+void GroupedStatsRequest::renderStat(const std::string& name, Buffer::Instance& response,
+                                     const StatOrScopes& variant) {
+  auto prefixed_tag_extracted_name = prefixedTagExtractedName<SharedStatType>(name);
+  if (prefixed_tag_extracted_name.has_value()) {
+    PrometheusStatsRender* const prometheus_render =
+        dynamic_cast<PrometheusStatsRender*>(render_.get());
+    // increment stats count
+    ++phase_stat_count_;
+
+    // sort group
+    std::vector<SharedStatType> group = absl::get<std::vector<SharedStatType>>(variant);
+    std::sort(group.begin(), group.end(),
+              [this](const Stats::RefcountPtr<Stats::Metric>& stat1,
+                     const Stats::RefcountPtr<Stats::Metric>& stat2) -> bool {
+                return global_symbol_table_.lessThan(stat1->statName(), stat2->statName());
+              });
+
+    // render group
+    StatOrScopesIndex index = static_cast<StatOrScopesIndex>(variant.index());
+    std::string type = (index == StatOrScopesIndex::Counter) ? "counter" : "gauge";
+
+    response.add(fmt::format("# TYPE {0} {1}\n", prefixed_tag_extracted_name.value(), type));
+    for (SharedStatType metric : group) {
+      prometheus_render->generate(response, prefixed_tag_extracted_name.value(), *metric.get());
+    }
+  }
+}
+
+void GroupedStatsRequest::processTextReadout(const std::string& name, Buffer::Instance& response,
+                                             const StatOrScopes& variant) {
+  renderStat<Stats::TextReadoutSharedPtr>(name, response, variant);
+}
+
+void GroupedStatsRequest::processCounter(const std::string& name, Buffer::Instance& response,
+                                         const StatOrScopes& variant) {
+  renderStat<Stats::CounterSharedPtr>(name, response, variant);
+}
+
+void GroupedStatsRequest::processGauge(const std::string& name, Buffer::Instance& response,
+                                       const StatOrScopes& variant) {
+  renderStat<Stats::GaugeSharedPtr>(name, response, variant);
+}
+
+void GroupedStatsRequest::processHistogram(const std::string& name, Buffer::Instance& response,
+                                           const StatOrScopes& variant) {
+  auto histogram = absl::get<std::vector<Stats::HistogramSharedPtr>>(variant);
+  auto prefixed_tag_extracted_name = prefixedTagExtractedName<Stats::HistogramSharedPtr>(name);
+
+  if (prefixed_tag_extracted_name.has_value()) {
+    // increment stats count
+    phase_stat_count_++;
+
+    // sort group
+    std::sort(histogram.begin(), histogram.end(),
+              [this](const Stats::RefcountPtr<Stats::Metric>& stat1,
+                     const Stats::RefcountPtr<Stats::Metric>& stat2) -> bool {
+                return global_symbol_table_.lessThan(stat1->statName(), stat2->statName());
+              });
+
+    // render group
+    response.add(fmt::format("# TYPE {0} {1}\n", prefixed_tag_extracted_name.value(), "histogram"));
```

Comment:
I guess this is where a Render interface might be useful. Let's just call this out as prometheus-specific and it could be refactored incrementally later.

Also, you *do* have a file of prom rendering functions -- should we at least move this magic there?